pulley/README.md

```diff
@@ -47,18 +47,14 @@ to change, instructions to appear and disappear, and APIs to be overhauled.
 Here is the disassembly of `f(a, b) = a + b` in Pulley today:
 
 ```
-       0: 0e 1f f0                        xconst8 x31, -16
-       3: 12 20 20 1f                     xadd32 sp, sp, x31
-       7: 29 20 08 21                     store64_offset8 sp, 8, lr
-       b: 27 20 22                        store64 sp, fp
-       e: 0b 22 20                        xmov fp, sp
-      11: 12 00 00 01                     xadd32 x0, x0, x1
-      15: 0b 20 22                        xmov sp, fp
-      18: 25 21 20 08                     load64_offset8 lr, sp, 8
-      1c: 22 22 20                        load64 fp, sp
-      1f: 0e 1f 10                        xconst8 x31, 16
-      22: 12 20 20 1f                     xadd32 sp, sp, x31
-      26: 00                              ret
+       0: 2d 01                           spush lr
+       2: 2d 02                           spush fp
```

Comment:
I was indeed thinking that the special registers would only be accessed from special instructions, but I think we can do even more here by bundling the whole function prologue and epilogue up into instructions that just take the size to decrement the stack frame by:

```
enter_frame 16
...
exit_frame 16
ret
```

where

```
enter_frame N

==>

push lr
push fp
fp = sp
sp = sp - N
```

and the opposite for `exit_frame N`.

Reply:
I'm a bit confused about the purpose of the `N` to `enter_frame`/`exit_frame`. Each `push`/`pop` will itself automatically adjust the sp by `size_of<*u8>()`. So in the case of no other registers being stored on the stack, the argument will be zero. In the case where other registers need to be saved on the stack, `xpush{32,64}` and `xpop{32,64}` will also adjust the sp by 4/8 bytes. So in either case the argument is not needed? I don't think there is any scenario where you would need to adjust the stack pointer without also reading/writing from/to the stack

Reply:
> I'm a bit confused about the purpose of the `N` to `enter_frame`/`exit_frame`.

To allocate the stack frame for this function, including all of its regalloc spills and all that.

That said, the way that Cranelift is organized and the abstraction boundaries it has might make doing the full function prologue in a single instruction difficult. We might have to settle for two instructions:

1. a macro instruction for `push lr; push fp; fp = sp` (and the opposite for exiting the function, maybe just fold this into `ret`)
2. an instruction for allocating and deallocating stack space: `sp = sp +/- imm32`

See `gen_prologue_frame_setup`, `gen_epilogue_frame_restore`, and `gen_sp_reg_adjust` in `../pulley_shared/abi.rs` in https://github.com/bytecodealliance/wasmtime/pull/9089 (or the equivalent methods in, e.g., `../aarch64/abi.rs`). I'm thinking that each of these should produce a single macro instruction because they otherwise always generate the exact same sequence of instructions modulo immediate constant values.

The benefit of a single macro instruction, compared to a sequence of composable and more general instructions, is performance. The expensive part of an interpreter is roundtripping through the opcode loop, by using macro instructions we are amortizing that cost.

Does that make sense?

Reply:
Decided the extra argument was not needed after all:
https://bytecodealliance.zulipchat.com/#narrow/stream/217126-wasmtime/topic/Pulley.20registers/near/459533622